0000-event-system.md

```diff
@@ -0,0 +1,109 @@
+- Start Date: 2018-01-16
+- RFC PR: (leave this empty)
+- React Navigation Issue: (leave this empty)
+
+# Summary
+
+Add an event emitter to the navigation prop that emits focus/blur events for the route.
+
+This proposal has been pioneered by @satya164: https://github.com/react-navigation/react-navigation/issues/1363 https://github.com/react-navigation/react-navigation/issues/51
+
+# Basic example
+
+In this example, we focus the first text input when the screen has completed transitioning:
+
+```
+componentDidMount() {
+	this._sub = this.props.navigation.events.addListener(
+		'DidFocus',
+		this._focusFirstTextInput
+	);
+}
+componentWillUnmount() {
+	this._sub.remove();
+}
+```
+
+
+# Motivation
+
+Screens need to know about navigation changes while they are mounted. Consider the following use cases:
+
+### Data and Media subscription
+Screens must know when to request live data or resource-intensive APIs. Media should be able to auto-pause at correct times.
+
+### Unmounting
+Tabs and cards may need to unmount and re-mount contents when they are focused.
+
+### Back button handling
+Screens may want to prevent back actions based on internal state
```

Comment:
I think it's rare for this to actually happen on iOS, it's more of an Android concern because you can press a back button at any time. On iOS if the back button is visible and not greyed out, and you can swipe back, then you should be able to go back. We should call out Android specifically here

Reply:
Ok I can definitely call out Android here. But I would note that this Android feature has a similarity on iOS: the back gesture being enabled or not. Right now we have an independent mechanism to control that, but it is super painful to use when it depends on internal screen state.

Reply:
oh interesting! thanks for pointing that out. we should mention it here too